### ISSUE TITLE
[FEATURE] Add support for Flux<String> in addition to TokenStream in AI service interfaces

### ISSUE THREAD

**Issue by marcushellberg** [langchain4j/langchain4j#886]
**Is your feature request related to a problem? Please describe.**
I typically stream responses. Because I'm in a Spring Boot environment, it's more convenient for me to work with a Flux data type than a TokenStream.

**Describe the solution you'd like**
Ideally, I could either define the method in the interface to return a Flux, or that I could call `tokenStream.asFlux()` 

**Describe alternatives you've considered**
I now manually create a Sink and emit next/complete/error on it to get a Flux.

**Comment by langchain4j** [comment]
[Agreed](https://github.com/langchain4j/langchain4j-spring/blob/6b02420b7b6e960afb40f19af1844347a13c3c50/langchain4j-spring-boot-starter/src/main/java/dev/langchain4j/service/spring/AiService.java#L100) :)

**Comment by yongxingMa** [comment]
How to use tokenstream in springboot environment？Now print to data on the console。
How to return streaming data to an API service or frontservice？
thank you！

**Comment by yongxingMa** [comment]
@GetMapping("/assistant")
    public TokenStream assistant(@RequestParam(value = "message", defaultValue = "hello") String message) {
        TokenStream tokenStream = assistant.chat(message);
        tokenStream.onNext(System.out::println)
                .onComplete(System.out::println)
                .onError(Throwable::printStackTrace)
                .start();
        return tokenStream;
    }

**Comment by yongxingMa** [comment]
```
@GetMapping("/assistant")
public TokenStream assistant(@RequestParam(value = "message", defaultValue = "hello") String message) {
        TokenStream tokenStream = assistant.chat(message);
        tokenStream.onNext(System.out::println)
                .onComplete(System.out::println)
                .onError(Throwable::printStackTrace)
                .start();
        return tokenStream;
    }

**Comment by langchain4j** [comment]
@yongxingMa here is an [example](https://github.com/marcushellberg/java-ai-playground/blob/e983c855eccd9c070a8e3fb11426609abb1ac335/src/main/java/org/vaadin/marcus/client/AssistantService.java#L35) of converting `TokenStream` into a `Flux<String>`.

**Comment by yongxingMa** [comment]
I'm used Flux<String> and I'm found that it returns FE not streaming data, and I'd want to use a request like SSE , is there any way to do that? Thank you!
```
public Flux<String> assistantFlux(String message) {
        assistant = AiServices.builder(Assistant.class)
                .streamingChatLanguageModel(zhiPuAi.getZhiPuAiStreamChatModel())
                .tools(new AssistantToolExecution())
                .chatMemory(MessageWindowChatMemory.withMaxMessages(10))
                .build();
        Sinks.Many<String> sink = Sinks.many().unicast().onBackpressureBuffer();
        assistant.chatWithStream(message)
                .onNext(sink::tryEmitNext)
                .onComplete(aiMessageResponse -> sink.tryEmitComplete())
                .onError(sink::tryEmitError)
                .start();
        return sink.asFlux();
    } 
```
Console Output：
```
2024-04-15T11:02:43.249+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onOpen()
2024-04-15T11:02:43.249+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":"I"}}]}
2024-04-15T11:02:43.250+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":"'m"}}]}
2024-04-15T11:02:43.250+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" good"}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":","}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" thank"}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" you"}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":"!"}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" How"}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" can"}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" I"}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" assist"}}]}
2024-04-15T11:02:43.251+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" you"}}]}
2024-04-15T11:02:43.252+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" today"}}]}
2024-04-15T11:02:43.259+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":"?"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" If"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" you"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" have"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" any"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" questions"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" or"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" need"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" help"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" with"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" something"}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":","}}]}
2024-04-15T11:02:43.260+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" feel"}}]}
2024-04-15T11:02:43.261+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" free"}}]}
2024-04-15T11:02:43.261+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" to"}}]}
2024-04-15T11:02:43.261+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":" ask"}}]}
2024-04-15T11:02:43.261+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"delta":{"role":"assistant","content":"."}}]}
2024-04-15T11:02:43.261+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() {"id":"8561163774422257761","created":1713150161,"model":"glm-4","choices":[{"index":0,"finish_reason":"stop","delta":{"role":"assistant","content":""}}],"usage":{"prompt_tokens":35,"completion_tokens":32,"total_tokens":67}}
2024-04-15T11:02:43.261+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onEvent() [DONE]
2024-04-15T11:02:43.261+08:00 DEBUG 19733 --- [bigmodel.cn/...] d.langchain4j.model.zhipu.ZhipuAiClient  : onClosed()
```

PostMan Output：
```
I'm good, thank you! How can I assist you today? If you have any questions or need help with something, feel free to ask.
```

**Comment by yongxingMa** [comment]
Used the following code solved my problem，Modify type 
MediaType.TEXT_EVENT_STREAM_VALUE
```
    @PostMapping(path = "/assistantFlux", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<String> assistantFlux(@RequestBody AssistantDto assistantDto) {
          assistant = AiServices.builder(Assistant.class)
                .streamingChatLanguageModel(zhiPuAi.getZhiPuAiStreamChatModel())
                .tools(new AssistantToolExecution())
                .chatMemory(MessageWindowChatMemory.withMaxMessages(10))
                .build();

        Sinks.Many<String> sink = Sinks.many().unicast().onBackpressureBuffer();
        assistant.chatWithStream(message)
                .onNext(sink::tryEmitNext)
                .onComplete(aiMessageResponse -> sink.tryEmitComplete())
                .onError(sink::tryEmitError)
                .start();
        return sink.asFlux();
    }
```
<img width="1374" alt="image" src="https://github.com/langchain4j/langchain4j/assets/31605966/736c5e03-29d2-4e1b-930c-687679c3e0c2">

**Comment by langchain4j** [comment]
Related: https://github.com/langchain4j/langchain4j/discussions/322

**Comment by langchain4j** [comment]
Related: https://github.com/langchain4j/langchain4j/issues/689

**Comment by langchain4j** [comment]
Will be released in 0.35.0 in a few weeks

**Comment by langchain4j** [comment]
To use `Flux<String>` as a return type, please import `langchain4j-reactor` module:
```xml
<dependency>
    <groupId>dev.langchain4j</groupId>
    <artifactId>langchain4j-reactor</artifactId>
    <version>${langchain4j.version}</version>
</dependency>
```

```java
interface Assistant {

    Flux<String> chat(String userMessage);
}
```
